### PR TITLE
[4.1] Form list with showon attributes should load the show on script

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue #37299

### Summary of Changes

- Load the showon script if there are showon attributes even in the options (one level down)

### Testing Instructions

- Open file: `\joomla-cms\administrator\components\com_banners\forms\banner.xml`
- Find the state field and replace with the following

```xml
		<field
			name="state"
			type="list"
			label="JSTATUS"
			class="form-select-color-state"
			default="1"
			validate="options"
			showon="state!:X"
			>
			<option value="1" showon="sticky:0">JPUBLISHED</option>
			<option value="0" showon="sticky:0">JUNPUBLISHED</option>
			<option value="2" showon="sticky:0">JARCHIVED</option>
			<option value="-2" showon="sticky:0">JTRASHED</option>
		</field>
```
- Create a new banner: `/administrator/index.php?option=com_banners&view=banner&layout=edit`
- Toggle Banner's Pinned button. You'll see the Status options disappear after toggling. (You'll first have to click the status list in order to see it change. It would be better for it to update, but that's another issue)
- Now remove the `showon="state!:X"` line from the code above. Press CTRL+F5.
- Retry toggling, it's broken now.

### CODE REVIEW

File: libraries/src/Form/Field/ListField.php [hunk 1]
```diff
@@ -45,6 +45,14 @@ class ListField extends FormField
 	 */
 	protected $layout = 'joomla.form.field.list';
 
+	/**
+	 * Does the list has showOn
+	 *
+	 * @var    boolean
+	 * @since  4.1.1
```
Comment (richard67) [inline libraries/src/Form/Field/ListField.php]:
```suggestion
	 * Does the list have showOn
	 *
	 * @var    boolean
	 * @since  __DEPLOY_VERSION__
```

**Comment by toroworx** [comment]
I just tested this patch and can confirm it fixes the issue.
Thanks for creating a patch! Awesome feature btw.

**Comment by richard67** [comment]
@toroworx Could you go to the PR in the issue tracker here https://issues.joomla.org/tracker/joomla-cms/37300 and use the "Test this" button to submit your test result? Thanks in advance.

**Comment by toroworx** [comment]
> @toroworx Could you go to the PR in the issue tracker here https://issues.joomla.org/tracker/joomla-cms/37300 and use the "Test this" button to submit your test result? Thanks in advance.

I tried to, but after clicking the "Login with Github" botton and clicking "Authorize Joomla" I'll get a 500 error:
```
Looks like something went wrong!

We track these errors automatically, but if the problem persists feel free to contact us. In the meantime, try refreshing.
Authorize Joomla Issue Tracker
```

**Comment by toroworx** [comment]
I have tested this item :white_check_mark: successfully on ac8bed219fa2e5848bfcc28f4190b168c3147aad<br /><br />I just tested this patch and can confirm it fixes the issue. After the patch you don't need to manually load the showon.min.js script anymore.

I just found out though: to replicate the issue, you have to make sure there's not a field with a showon attribute, else this will load the showon.min.js script, preventing you from replicating the issue. Nevertheless the patch works fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37300">issues.joomla.org/tracker/joomla-cms/37300</a>.</sub>

File: layouts/joomla/form/field/list.php [hunk 2]
```diff
@@ -42,9 +43,10 @@
  * @var   array    $options         Options available for this field.
  * @var   string   $dataAttribute   Miscellaneous data attributes preprocessed for HTML output
  * @var   array    $dataAttributes  Miscellaneous data attribute for eg, data-*
+ * @var   boolen   $hasShowOn       Is the list using the showon
```
Comment (Quy) [inline layouts/joomla/form/field/list.php:46]:
```suggestion
 * @var   Boolean  $hasShowOn       Is the list using the showon
```

**Comment by dgrammatiko** [comment]
I have changed this PR so it will use the Platform wide `$options['showonEnabled']` https://github.com/joomla/joomla-cms/blob/29ec453f5b36672fd530618eeadd8898e806c572/libraries/src/Form/FormField.php#L1070-L1075 which is held automatically here https://github.com/joomla/joomla-cms/blob/29ec453f5b36672fd530618eeadd8898e806c572/layouts/joomla/form/renderfield.php#L28-L33

**Comment by richard67** [comment]
@toroworx Could you test again with the latest changes? Thanks in advance.

**Comment by toroworx** [comment]
Seems to be broken right now. 

### Test instructions:
1. Open file: \joomla-cms\administrator\components\com_banners\forms\banner.xml
2. Find the state field and replace with the following
```
		<field
			name="state"
			type="list"
			label="JSTATUS"
			class="form-select-color-state"
			default="1"
			validate="options"
			showon="state!:X"
			>
			<option value="1" showon="sticky:0">JPUBLISHED</option>
			<option value="0" showon="sticky:0">JUNPUBLISHED</option>
			<option value="2" showon="sticky:0">JARCHIVED</option>
			<option value="-2" showon="sticky:0">JTRASHED</option>
		</field>
```
3. Create a new banner: /administrator/index.php?option=com_banners&view=banner&layout=edit
4. Toggle Banner's `Pinned` button. You'll see the `Status` options disappear after toggling. 
(You'll first have to click the status list in order to see it change. It would be better for it to update, but that's another issue)
5. Now remove the `showon="state!:X"` line from the code above. Press CTRL+F5.
5. Retry toggling, it's broken now.

**Comment by richard67** [comment]
@toroworx Have you applied the changes in ListField.php AND also reverted the List.php to the original state like without this PR? Or have you maybe forgotten the latter?

**Comment by laoneo** [comment]
@toroworx what do you expect instead of that the status are hidden? For me it works exact the same way as you did describe it in the last comment which is for me the expected behavior.

**Comment by toroworx** [comment]
@richard67 I've used the Joomla Patch Tester. I reverted the patch, fetched data, and applied the patch.
Just to be sure I undid all that. Switched to the patch-17 branch, pulled in the latest version and retested (so testing 723f805a7eba122353866ce77fa6d05ed52120a5). Again I can confirm the latest patch-17 version is broken.

@laoneo did you refresh after step 5? The definition of broken is: not hiding the `Status` options any more after toggling the banner's `Pinned` button (you have to click the Status option to view if the options are hidden). To clarify, this only happens if _only option showon is used_. So like this:

```
<field
		name="state"
		type="list"
		label="JSTATUS"
		class="form-select-color-state"
		default="1"
		validate="options"
>
	<option value="1" showon="sticky:0">JPUBLISHED</option>
	<option value="0" showon="sticky:0">JUNPUBLISHED</option>
	<option value="2" showon="sticky:0">JARCHIVED</option>
	<option value="-2" showon="sticky:0">JTRASHED</option>
</field>
```

If on the page somewhere `showon` is used within a `field` tag, the showon is loaded and the issue can't be replicated.

**Comment by dgrammatiko** [comment]
@toroworx you're right the latest patch is not working as expected

@laoneo there's problem with the List field (which is also the base for many other fields). The showon is loaded only if the field is declaring a showon attribute in the root, if it is in a nested option the showon is not loaded. This affects **many** other fields. My first approach was naive, eg have a new option passed in the layout data but I realised that will not work for any of the fields that are extending the List field. Also I have no clue how this can be achieved without some refactoring (eg the getLayoutData or the getOptions fn)

**Comment by laoneo** [comment]
Thanks for clarification, now I got it.

File: libraries/src/Form/FormField.php [hunk 3]
```diff
@@ -1360,6 +1360,7 @@ protected function getLayoutData()
 				if ((string) $option['showon'])
 				{
 					$options['showonEnabled'] = true;
+					break;
```
Comment (richard67) [inline libraries/src/Form/FormField.php:1364]:
```suggestion

					break;
```

**Comment by laoneo** [comment]
@dgrammatiko did your last changes fix the bug which was found by @toroworx?

**Comment by dgrammatiko** [comment]
@laoneo it should solve the reported issue, if not then it should be closed

**Comment by toroworx** [comment]
/edited:
I just pulled the latest commits of your patch-17 branch and tested it. It doesn't fix it. Still same state as before the patch. showon.min.js is not loaded. It's still acting like [explained above](https://github.com/joomla/joomla-cms/pull/37300#issuecomment-1073211308).

I appreciate your pull request, but why do you suggest it's fixed before testing it? With the provided [instructions](https://github.com/joomla/joomla-cms/pull/37300#issuecomment-1073211308) it shouldn't take long.

**Comment by dgrammatiko** [comment]
> but why do you suggest it's fixed before testing it? With the provided https://github.com/joomla/joomla-cms/pull/37300#issuecomment-1073211308 it shouldn't take long.

I did and the showon file was loaded. Anyways someone else could have better lack with this

**Comment by toroworx** [comment]
> I did and the showon file was loaded. Anyways someone else could have better lack with this

Okay, I'm sorry then. Weird it did work for you and not for me.

@dgrammatiko I've edited your approach and this works for me now. Can you please review the approach?

https://github.com/toroworx/joomla-cms/commit/d428b1ffb2e35e787dc7f3a08fd7becafc60f045

**Comment by richard67** [comment]
I‘ve re-opened the issue. But @dgrammatiko could you check the commit linked by @toroworx ?

**Comment by richard67** [comment]
Meanwhile @toroworx has made a PR, see #37451 .

**Comment by toroworx** [comment]
> Meanwhile @toroworx has made a PR, see #37451 .

Thanks for reopening @richard67. I don't want to hijack anything, if I'll need to close something please let me know!

**Pull request closed without merging**